WalletWasabi/BestEffortEndpointConnector.cs

```diff
@@ -0,0 +1,89 @@
+using NBitcoin.Protocol;
+using NBitcoin.Protocol.Behaviors;
+using NBitcoin.Protocol.Connectors;
+using System;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace WalletWasabi
+{
+	public class BestEffortEndpointConnector : IEnpointConnector
+	{
+        public enum ConnectionMode
+        {
+            // Connect to nodes published as onion services only 
+            OnionServiceOnly,
+            // Connect to nodes published on the clearnet through Tor exit nodes
+            AllowGoingThroughTorExitNodes,
+            // Connect to nodes directly on clearnet
+            ClearNet
+        }
+
+		public class EffortState
+        {
+            public ConnectionMode Mode { get; set; }
+            public DateTimeOffset LastConnectionTime { get; set; }
+            public long ConnectedNodesCount { get; set; }
+
+            public EffortState(ConnectionMode mode, DateTimeOffset lastConnectionTime)
+            {
+                Mode = mode;
+                LastConnectionTime = lastConnectionTime;
+            }
+        }
+
+        public DefaultEndpointConnector Connector { get; private set; }
+        public EffortState State { get; private set; }
+
+        public BestEffortEndpointConnector()
+            : this(
+                new DefaultEndpointConnector(allowOnlyTorEndpoints: true), 
+                new EffortState(ConnectionMode.OnionServiceOnly, DateTimeOffset.UtcNow))
+        {
+        }
+
+        private BestEffortEndpointConnector(DefaultEndpointConnector connector, EffortState state)
+        {
+            Connector = connector;
+            State = state;
+        }
+
+		public void IncreaseConnectedNodesCounter()
+		{
+			State.ConnectedNodesCount++;
+		}
+
+		public IEnpointConnector Clone()
+		{
+			return new BestEffortEndpointConnector((DefaultEndpointConnector)Connector.Clone(), State);
+		}
+
+		public virtual async Task ConnectSocket(Socket socket, EndPoint endPoint, NodeConnectionParameters nodeConnectionParameters, CancellationToken cancellationToken)
+		{
+            var elapsedTimeSinceLastConnection = DateTimeOffset.UtcNow - State.LastConnectionTime;
+
+            if (elapsedTimeSinceLastConnection > TimeSpan.FromMinutes(3) && State.ConnectedNodesCount == 0)
+            {
+                switch (State.Mode)
```

Comment:
Logging statements with appropriate levels may be useful for real-time debugging / behavior observing.